src/CineControl.CinemaService.API/Service/CinemaService.cs

```diff
@@ -0,0 +1,248 @@
+using CineControl.CinemaService.API.Data;
+using CineControl.CinemaService.API.Errors;
+using CineControl.CinemaService.API.Models;
+using CineControl.CinemaService.API.Models.DTOs.Cinemas;
+using CineControl.CinemaService.API.Models.DTOs.Theaters;
+using CineControl.CinemaService.API.Service.IService;
+using CineControl.Common.Results;
+using CineControl.Common.Tenant;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+
+namespace CineControl.CinemaService.API.Service
+{
+    public class CinemaService : ICinemaService
+    {
+        private readonly CinemaContext _context;
+        private readonly ITenantProvider _tenantProvider;
+
+        public CinemaService(CinemaContext context, ITenantProvider tenantProvider)
+        {
+            _context = context;
+            _tenantProvider = tenantProvider;
+        }
+
+        public async Task<ResultT<IEnumerable<Cinema>>> GetAllCinemas()
+        {
+            var cinemas = await _context.Cinemas
+                .Where(c => c.TenantId == _tenantProvider.TenantId)
+                .Include(c => c.Theaters)
+                .ThenInclude(t => t.Seats)
+                .ToListAsync();
+            return cinemas;
+        }
+
+        public async Task<ResultT<Cinema>> GetCinemaById(int id)
+        {
+            var cinema = await _context.Cinemas
+                .Where(c => c.TenantId == _tenantProvider.TenantId && c.Id == id)
+                .Include(c => c.Theaters)
+                .ThenInclude(t => t.Seats)
+                .FirstOrDefaultAsync();
+
+            return cinema is not null
+                ? cinema
+                : CinemaErrors.NotFound($"Cinema with id {id} not found");
+        }
+
+        public async Task<ResultT<IEnumerable<string>>> GetAllCities()
+        {
+            var cities = await _context.Cinemas
+                .Where(c => c.TenantId == _tenantProvider.TenantId)
+                .Select(c => c.City)
+                .Distinct()
+                .ToListAsync();
+
+            return cities;
+        }
+
+        public async Task<ResultT<IEnumerable<Cinema>>> GetCinemasByCity(string city)
+        {
+            var cinemas = await _context.Cinemas
+                .Where(c => c.TenantId == _tenantProvider.TenantId && c.City.ToLower() == city.ToLower())
+                .ToListAsync();
+
+            return cinemas;
+        }
+
+        public async Task<ResultT<Cinema>> AddCinema(AddCinemaRequest request)
+        {
+            if (!_tenantProvider.HasTenant)
+            {
+                return CinemaErrors.AccessUnauthorized("No tenant specified");
```

Comment:
to najlepiej też to `CinemaError` i jak już rozróżnić to osobny typ erroru np `CinemaError.TenantAccessUnauthorized` i który tworzy błąd `Error.AccessUnauthorized`